lib/Backends/CPU/CPUFunction.cpp

```diff
@@ -13,28 +13,80 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 #include "CPUFunction.h"
 
+#include "glow/Graph/Context.h"
 #include "glow/Support/Compiler.h"
 #include "glow/Support/Memory.h"
 
 using namespace glow;
 
-CPUFunction::CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT, void *heap)
-    : JIT_(std::move(JIT)), heap_(heap) {}
+CPUFunction::CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT,
+                         const runtime::RuntimeBundle &runtimeBundle)
+    : JIT_(std::move(JIT)), runtimeBundle_(runtimeBundle) {}
+
+CPUFunction::~CPUFunction() { alignedFree(runtimeBundle_.constants); }
 
-CPUFunction::~CPUFunction() { alignedFree(heap_); }
+void CPUFunction::allocateMutableBuffersOnDevice() {
+  baseActivationsAddress_ = (uint8_t *)alignedAlloc(
+      runtimeBundle_.activationsMemSize, TensorAlignment);
+
+  baseMutableWeightVarsAddress_ = (uint8_t *)alignedAlloc(
+      runtimeBundle_.mutableWeightVarsMemSize, TensorAlignment);
+}
+
+void CPUFunction::copyMutablesToDevice(Context &ctx) {
+  // Copy Placeholders into allocated memory.
+  for (auto PH : ctx.pairs()) {
```

Comment:
placeholders contain both inputs and outputs. Do we need to copy outputs? What's your plan here?

Reply:
I don't know if we currently have a good way to filter this from the context. For this PR I suggest we copy all. I've created a separate issue #1953 to track how we'd like to pass inputs/outputs going forward.